Change fallocate() failures from WARNING to FATAL

vjr opened the pull request:
Fixes #3871

vjr commented:
Running the tests with `yb_build --ctests` results in the following conclusion:
```97% tests passed, 12 tests failed out of 352

Total Test time (real) = 3741.89 sec

The following tests FAILED:
     61 - util_background_task-test (Failed)
     93 - util_monotime-test (Failed)
    182 - rpc_rpc-test (Failed)
    183 - rpc_rpc_stub-test (Failed)
    230 - client_snapshot-schedule-test (Failed)
    261 - integration-tests_remote_bootstrap-itest (Failed)
    281 - integration-tests_cassandra_cpp_driver-test (Failed)
    284 - integration-tests_cql-tablet-split-test (Failed)
    291 - integration-tests_cdc_service-txn-test (Failed)
    293 - integration-tests_twodc-test (Failed)
    346 - pgwrapper_pg_libpq-test (Failed)
    348 - pgwrapper_pg_mini-test (Failed)
Errors while running CTest
```
I will check the logs later on, but wanted to know if these are known failures?

vjr commented:
The issue #3871 comment by @bmatican asks to change from WARNING to FATAL which I did, but I'm unsure about the second part to perform a check earlier in main/init for `fallocate()` failure without actually calling the function.

While reproducing the issue using `yugabyted` and `yb-ctl` tools the startup fails early and fast enough it seems so this change should suffice?

vjr commented:
I'm getting many test failures upon running `./yb_build.sh --ctest release` on both git master and my PR branch, on the order of 100+ failures.

Marking this PR as ready for review if someone else would like to take a look.

tedyu commented on src/yb/rocksdb/util/io_posix.cc:
```diff
@@ -474,11 +486,18 @@ Status PosixWritableFile::Allocate(uint64_t offset, uint64_t len) {
-  } else {
-    return STATUS_IO_ERROR(filename_, errno);
+
+  if (alloc_status != 0) {
```
Same check is repeated in the PR.

Can you extract the check into a helper method ?

vjr replied:
Done! Hopefully the modification to headers/includes and the name of the helper method is not too bad!

tedyu commented:
Can you look at the test failures and see which ones were caused by the new FATAL logs ?

Please list a few such tests - I will run them myself.

Thanks

vjr commented:
> I ran some test with the patch which passed:
> 
> ./yb_build.sh --cxx-test integration-tests_tablet-split-itest --gtest_filter TabletSplitITest.TestLoadBalancerAndSplit
> 
> Can you look at the test failures and see which ones were caused by the new FATAL logs ?
> 
> Please list a few such tests - I will run them myself.
> 
> What OS do you run the tests with ?
> 
> Thanks

Thanks for your inputs! I built and ran the tests on stock Ubuntu 20.04 LTS and I have the ulimits changes done.

If you see my earlier comment at that time a handful of tests failed: see https://github.com/yugabyte/yugabyte-db/pull/8569#issuecomment-846478326

But when I again updated from git latest changes from master I'm getting a lot of failures (over 100 tests).

Can you point me to where I can locate the failure logs to look at? I tried grep the test log folder but I only see lines that have `failures="0"` in them, can't seem to locate the failure logs.

In the mean time I will keep looking and rerun the tests after addressing your review comment about avoiding the repetitive code.

tedyu commented:
You can use commit 2ca94c4e2a55bf7fab5483c2215dbe4f09c21f17 . There have been some recently introduced test failures at the top of master branch.

I normally run single test e.g.:

./yb_build.sh --java-test 'org.yb.loadtest.TestSpark3Jsonb#testCdr' > /tmp/cdr.out 2>&1 &

In cdr.out you should see log similar to the following:
```
[postprocess_test_result.py:145] 2021-05-27 01:46:43,364 INFO: Log path: /nfusr/dev-server/tedyu/yugabyte-db/java/yb-cql-4x/target/surefire-reports_org.yb.loadtest.TestSpark3Jsonb__testCdr/org.yb.loadtest.TestSpark3Jsonb-output.txt
```
You can examine the test output in that file.

Once you update your PR, I can submit to YB-internal Jenkins so that we know which test(s) fail (due to your patch) (hopefully no new test failure).

vjr commented:
clang failure may require change `LOG_OCCURRENCES` from `int` to `uint64_t` in all macros except one which already has it as `uint64_t` in `src/yb/util/logging.h` - I will try to build locally with clang failing which will just push the change after building with gcc to let the CI do the check for me.

tedyu commented on src/yb/util/errno.h:
```diff
@@ -47,6 +48,18 @@ inline std::string ErrnoToString(int err) {
+    } else if (err == ENOSYS) {
+      YB_LOG_FIRST_N(FATAL, 1) << "The kernel does not implement fallocate().";
+    } else {
+      YB_LOG_FIRST_N(FATAL, 1) << "fallocate() failed.";
```
In the current code, STATUS_IO_ERROR() is returned.
Can LogFatalForFallocateFailure() return Status for this case (which is then returned in the caller) ?

vjr replied:
Yes, an error status was being returned and I believe that is also what was intended to be changed to fatal so that the YB scripts would fail fast, even in the case of unhandled errno from fallocate() failure. Do we really want to put that status error back?

tedyu commented:
Looking at the issue, it seems the following should be done as well:
```
add this check to the very top-level (in the init phase in main)
```

vjr commented:
> Looking at the issue, it seems the following should be done as well:
> 
> ```
> add this check to the very top-level (in the init phase in main)
> ```

Yes, in my earlier comment here https://github.com/yugabyte/yugabyte-db/pull/8569#issuecomment-846478748 I had asked about this and mentioned that the YB scripts are failing fast anyway in my testing on NFS share with fallocate() support lacking, so wan't sure if this is needed.

I don't know a way to check for fallocate() support at runtime without actually calling the function itself, but that would require some known/configured filesystem path. Not sure if there is another way such as the `advise()` function. Do you have any suggestions?

tedyu commented:
I am polling my co-workers who are more familiar with fallocate().
Stay tuned.

vjr commented:
Marking this PR as draft again for now because I'm hitting issue https://github.com/yugabyte/yugabyte-db/issues/7735

Will try either building on CentOS install or working around build on ubuntu with clang before pushing further commits.

tedyu commented:
I am still waiting for my co-workers for their opinion.

Also, there are several known failed tests - it would be better to wait for Jenkins to get greener :-)

vjr commented:
Sure thing, no rush. Thanks for your time on this issue.

vjr commented:
> I am still waiting for my co-workers for their opinion.
> 
> Also, there are several known failed tests - it would be better to wait for Jenkins to get greener :-)

Hello @tedyu  would it be a good time to revisit this PR?

tedyu commented:
Haven't heard of anything from my coworkers.

@bmatican What's your suggestion for this PR ?

vjr commented:
@tedyu i am unable to do so myself so would you add @bmatican as a reviewer so he might look into it when he can? Thanks!

vjr commented:
Hello @bmatican - have you been able to look into this PR yet? I'll be glad to work on any review comments you or your team members might have for me. Thanks!

vjr commented:
I just read on YB latest docs [deployment checklist](https://docs.yugabyte.com/latest/deploy/checklist/) that NFS is not supported and is on the roadmap https://github.com/yugabyte/yugabyte-db/issues/4388

Just to note that in my testing YBDB appears to startup fine with NFSv4 which supports fallocate() while running in NFSv3 mode which lacks fallocate() capability then the issue occurs. I have not done further tests on NFS for functionality, this PR is just to fail fast upon attempt to run on NFS that lacks fallocate() capability.
